server/src/main/java/org/opensearch/cluster/routing/RoutingNode.java

```diff
@@ -48,63 +48,140 @@
 
+    static class BucketedShards implements Iterable<ShardRouting> {
+        private static Map<Boolean, Integer> map = new HashMap<Boolean, Integer>() {
+            {
+                put(true, 0);
+                put(false, 1);
+            }
+        };
+
+        private final LinkedHashMap<ShardId, ShardRouting>[] shards; // LinkedHashMap to preserve order
+
+        BucketedShards(LinkedHashMap<ShardId, ShardRouting> primaryShards, LinkedHashMap<ShardId, ShardRouting> replicaShards) {
+            this.shards = new LinkedHashMap[2];
+            this.shards[0] = primaryShards;
+            this.shards[1] = replicaShards;
+        }
+
+        public boolean isEmpty() {
+            return this.shards[0].isEmpty() && this.shards[1].isEmpty();
+        }
+
+        public int size() {
+            return this.shards[0].size() + this.shards[1].size();
+        }
+
+        public boolean containsKey(ShardId shardId) {
+            return this.shards[0].containsKey(shardId) || this.shards[1].containsKey(shardId);
+        }
+
+        public ShardRouting get(ShardId shardId) {
+            if (this.shards[0].containsKey(shardId)) {
+                return this.shards[0].get(shardId);
+            }
+            return this.shards[1].get(shardId);
+        }
+
+        public ShardRouting add(ShardRouting shardRouting) {
+            return put(shardRouting.shardId(), shardRouting);
+        }
+
+        public ShardRouting put(ShardId shardId, ShardRouting shardRouting) {
+            ShardRouting ret = this.shards[map.get(shardRouting.primary())].put(shardId, shardRouting);
+            if (this.shards[map.get(!shardRouting.primary())].containsKey(shardId)) {
+                return this.shards[map.get(!shardRouting.primary())].remove(shardId);
+            }
+
+            return ret;
+        }
+
+        public ShardRouting remove(ShardId shardId) {
+            if (this.shards[0].containsKey(shardId)) {
+                return this.shards[0].remove(shardId);
+            }
+            return this.shards[1].remove(shardId);
+        }
+
+        @Override
+        public Iterator<ShardRouting> iterator() {
+            final Iterator<ShardRouting> primaryIterator = Collections.unmodifiableCollection(shards[0].values()).iterator();
+            final Iterator<ShardRouting> replicaIterator = Collections.unmodifiableCollection(shards[1].values()).iterator();
+            return new Iterator<ShardRouting>() {
+                @Override
+                public boolean hasNext() {
+                    return primaryIterator.hasNext() || replicaIterator.hasNext();
+                }
+
+                @Override
+                public ShardRouting next() {
+                    if (primaryIterator.hasNext()) {
+                        return primaryIterator.next();
+                    }
+                    return replicaIterator.next();
+                }
+            };
+        }
```

Comment:
This iterator wouldn't allow replicas to tail primaries in the same iteration causing them to be strictly interleaved. The implication is shard recoveries will slow down, probably in usual cases.  A node is capable of recovering based on the node outgoing recoveries of the peer primaries. So its possible the outgoing on the peer primary still has capacity to accommodate an outgoing recovery, which the change might otherwise hinder

Reply:
@Bukhtawar - Thinking more on this, replicas tailing primaries probably won't impact the shard recovery speed as the corresponding peer primary shard won't have relocated initially. More generally, the probability of replica (assuming single replica case) relocating before its primary is .5. Hence, 50% of replicas are going to anyways recover from the source nodes. Assuming constant time for each shard relocation (x) and 100p/100r, we are going to take 150x time on average.
Whereas in the strict interleaving case, after 100x relocation time for primaries, the replicas can recover using corresponding primaries and should take similar time around 150x.

Similar analysis can be done for multiple replica scenario. For 2 replica scenario:
* Average time without this change for 100p, 200r is 200x - .25 * 100x (both replica relocate after primary), .5 * 200x (one replica before, one after), .25 * 300x (both replica relocate before primary)
* Average time with this change for 100p, 200r is ~200x again - 100x for primaries and ~100x for both the replicas

Reply:
While there could be other cases(need to think more) but one place where it might hurt is the long tail. In real world cases shards aren't same in size and hence we usually end up in cases where only few nodes, actively relocating gets to decide the worst case.
With node concurrent recoveries as 2 with shards moving out from A maybe we can squeeze in more, but will end up in more iterations due to interleaving
**Current**
```
4p/2r --> 2p/0r -->0p/0r [3 moves]
```
**Proposed**
```
4p/2r --> 2p/2r --> 0p/2r --> 0p/0r [4 moves]
```
Having said that I agree this might only be slightly worse in certain cases, wondering if you think placing the change behind the same setting makes sense? I'll let others weigh in too, though I don't see a major concern

Reply:
Thanks @Bukhtawar. Given this scenario is rare and not major concern as both the primaries need to finish recovery at same time, and that is even more unusual for real world scenarios (4-5 nodes, 8-10 concurrent recoveries), I am also not in favor of additional setting.